### PR TITLE
Insert type references into the database

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Previously, we only checked if a reference is valid, this PR also inserts it into the database

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202597697815314/f) _(internal)_

## 🚫 Blocked by

- #821 

## 🔍 What does this change?

- Because `sqlx::Executor` is only implemented for references, change the implementation to use `Transaction` instead
- Use transactions to read the references
- Store the references in the database

## ⚠️ Known issues

We have quite a few code duplication here as we need the logic for different types.

## 🐾 Next steps

- Add tests to the test suite, which passes invalid date: [Asana task](https://app.asana.com/0/1200211978612931/1202598311667201/f) (_internal_)

## 🛡 What tests cover this?

The integration test suite covers this

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs [hunk 1]
```diff
@@ -372,102 +360,184 @@ impl PostgresDatabase {
...
     {
-        futures::future::try_join_all(
-            property_type_references
-                .into_iter()
-                .map(|reference| self.version_id_by_uri(reference.uri())),
-        )
-        .await
+        let property_type_references = property_type_references.into_iter();
+        let mut ids = Vec::with_capacity(property_type_references.size_hint().0);
+        for reference in property_type_references {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:467]: I see that futures has been removed as a dependency as well, what is the reason for switching away from the `join_all`? 
Wouldn't this version be "slower" as we are awaiting at each element?

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs [hunk 2]
```diff
@@ -372,102 +360,184 @@ impl PostgresDatabase {
...
     {
-        futures::future::try_join_all(
-            data_type_references
-                .into_iter()
-                .map(|reference| self.version_id_by_uri(reference.uri())),
-        )
-        .await
+        let data_type_references = data_type_references.into_iter();
+        let mut ids = Vec::with_capacity(data_type_references.size_hint().0);
+        for reference in data_type_references {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:483]: Same comment here as for `property_type_reference_ids`

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs [hunk 3]
```diff
@@ -249,7 +249,7 @@ impl PostgresDatabase {
     ///
     /// [`BaseUri`]: crate::types::BaseUri
     async fn update<T>(
-        &self,
+        transaction: &mut Transaction<'_, Postgres>,
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:252]: I like that we can push the transactions one level up from here :+1:

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs [hunk 4]
```diff
@@ -372,102 +360,184 @@ impl PostgresDatabase {
     }
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:360]: (can't comment at the right location, but line 347, function `get_by_version`) If we're drive-by changing to the other query syntax, the function here is still using `fetch_one` with an executor. Should we change it here as well?

**Review by thehabbos007 — COMMENTED** [review]
Am aware that some of this might change when the other PR merges, but I left some comments about the current state

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:467]: It's because of the `Transaction` struct. It requires to be `&mut`, thus it can't be done in parallel.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs (hunk 4 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:363]: Good catch: 2b642138efd2ea5e92b21f21a7e836a07822d01e

**Comment by thehabbos007** [comment]
Just flagging that this might be blocked by https://github.com/hashintel/hash/pull/824

**Comment by TimDiekmann** [comment]
Given the simplicity of #824 I like to block that PR on this.

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts [hunk 5]
```diff
@@ -132,11 +132,13 @@ export async function up(pgm: MigrationBuilder): Promise<void> {
         type: "UUID",
         notNull: true,
         references: "property_types",
+        onDelete: "CASCADE",
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts:135]: @TimDiekmann you don't have to answer, I just wonder if anyone in @hashintel/backend has an idea

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs [hunk 6]
```diff
@@ -48,73 +48,85 @@ impl PostgresDatabase {
...
-                .report()
-                .change_context(QueryError)
-                .attach_printable_lazy(|| base_uri.clone())?;
-
-        Ok(exists)
+    async fn contains_base_uri(
+        transaction: &mut Transaction<'_, Postgres>,
+        base_uri: &BaseUri,
+    ) -> Result<bool, QueryError> {
+        Ok(transaction
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: It's only because of the poor implementation of the `Executor` trait. We need to use a `Transaction` here because we must not insert the type if any other inserts are failing.

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: My lack of past experience with this is holding me up a bit. Sorry for the work but can you use some more words and explain it in greater detail here? How does this transaction work with some of the other methods? Maybe it'd be helpful to loosely explain what happens with the transaction with`insert_property_type` or `update_property_type` or whatever.

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts:135]: We currently only have two options to properly run the tests: 
- Set up a database for each test, which has to be dropped after running the test
- Cleaning up the database after each test has run

We decided against the first solution for now for simplicity. The second solution requires a deletion method for inserted items. Originally, we don't want to expose the database, however, then we can't clean up the database for tests. Cascading is still easier than implementing a recursive deletion of all entries, as we only rely on two tables rather than on *all* tables.

Before we use the first solution, I don't see us removing the cascading.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: Getting rid of the lifetimes is pretty nice as well, but a minor thing

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts:135]: I think we should explicitly mention this then in the task about E2E tests. Because right now we are polluting our table schemas with dangerous unwanted functionality purely for the benefit of ergonomics of testing.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: Sure. To insert a property type, we want to check, if all references are valid. This is done in `insert_property_type_references`. That function will first read the `VersionId`s for the references. It will fail if a reference does not exist. The version ids are then used to store the reference inside the database. This implies, that the version id of the property type, which we want to insert, is already existing in the database, so inserting the actual type happens before.
Internally, inserting the `property_type` will use transactions (`fn create()`), as multiple things have to be inserted in order to store the blob. The scope of the transaction is now expanded to `insert_property_type_references`, so inserting the property type will fail if the references could not be stored.

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: I don't think I'm getting the information I need here 😁 lets go to a call

**Review by Alfred-Mountfield — APPROVED** [review]

File: packages/graph/migrations/postgres/migration/1656417312397_initial.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/migrations/postgres/migration/1656417312397_initial.ts:135]: Have added it.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:55]: We cleared that up :) and also, transactions are pain in SQLx
